Bug 1957668: show console URL when asking for Authentication

atiratree opened the pull request:
~There is no easy access for not logged in users to the console.~

~We have to guess the console URL, since we are not authenticated yet and we cannot be sure of the real URL (configurability docs: https://docs.openshift.com/container-platform/4.7/web_console/configuring-web-console.html)~

show console URL when asking for Authentication
- the URL will redirect to the real URL
- added check for a server version since this feature is not generally available before OpenShift 4.11 (Kubernetes 1.24)


The output of the command:

```bash
> oc login -u test
Authentication required for https://api.my-cluster0.com:6443 (openshift)
Console URL: https://api.my-cluster0.com:6443/console
Username: test
Password:

```

atiratree commented:
/bugzilla refresh

atiratree commented:
@soltysh 
I am not really sure if this is that good of a solution for such common use case, since we are guessing the URL.

atiratree commented:
/retest

soltysh commented on pkg/helpers/tokencmd/basicauth.go:
```diff
@@ -87,6 +89,10 @@ func (c *BasicChallengeHandler) HandleChallenge(requestURL string, headers http.
 		}
+		consoleURL, err := guessConsoleURL(c.Host)
+		if err == nil {
+			fmt.Fprintf(w, "Default console URL: %s\n", consoleURL)
```
Maybe say it's possibly at that URL, because we allow configuring that through https://github.com/openshift/api/blob/0a6cab75f657a12860a97faaf37ec71cd92530fa/kubecontrolplane/v1/types.go#L36

atiratree replied:
I am not sure if that is a good choice, ie `$API_URL/console` . Even though it can be set I have not found this documented anywhere. I am worried that for most of the users this will not work.

My solution is not great either and would prefer not to merge it. Because we are guessing the apps subdomain and the domain  of the cluster is configurable through ingress. So this will not work in many cases as well.

I think the best consistent solution would be for apiserver to detect a console URL by default if not specified , so the user would always get redirected to the right URL.

Thoughts?

atiratree replied:
ping @soltysh

soltysh replied:
@atiratree maybe we should have a hardcoded `$API_URL/console` which will redirect to appropriate console URL as configured. This would allow us to use that, but still would leave users with a functioning mechanism for exposing the URL where they wish. wdyt?

atiratree commented:
/bugzilla refresh

atiratree commented:
should be replaced by a new approach of redirecting to the console url by API server in https://github.com/openshift/kubernetes/pull/1110

/hold

soltysh reviewed (approved):
/lgtm
/approve
/bugzilla refresh
feel free to drop the hold once https://github.com/openshift/kubernetes/pull/1110 merges

atiratree commented:
this PR still has the older approach and needs to be rewritten to use https://github.com/openshift/kubernetes/pull/1110 so /hold till then

atiratree commented:
I have updated the implementation to point to the newly implemented api server console URL (`${API_SERVER}/console`).

I have also added version detection since this feature was not generally available before and could cause issues to the users who are connecting to older clusters.  Unfortunately  this can cause issues when connecting to vanilla 1.24+ Kubernetes.    

@soltysh please rereview

atiratree commented:
/hold cancel

atiratree commented:
/retest

atiratree commented:
/retest

atiratree commented:
/retest

atiratree commented:
/test e2e-metal-ipi-ovn-ipv6

atiratree commented:
/test e2e-metal-ipi-ovn-ipv6

atiratree commented:
/retest

soltysh reviewed (approved):
/lgtm
